cpp/src/arrow/python/extension_type.cc

```diff
@@ -193,5 +193,105 @@ Status UnregisterPyExtensionType() { return UnregisterExtensionType(kExtensionNa
 
 std::string PyExtensionName() { return kExtensionName; }
 
+
+// Define a generic extension type class that takes the name as constructor
+// parameter
+
+GenericExtensionType::GenericExtensionType(std::shared_ptr<DataType> storage_type,
```

Comment:
This (and below) is mostly the same as the code above for `PyExtensionType` with only a few changes (if we decide to keep both, I should look into deduplicating the code)

Reply:
Is it necessary to have the `GenericExtensionType` also? My initial thoughts were that `PyExtensionType` could be modified to use other names and use the registry

Reply:
There are some differences that I initially thought might be difficult to combine in a single class. But looking at it now again, I see the main difference is (apart from some checks of the constant name):

- `Serialize` always calls the instance (as it could be parametrized, so the serialization could be different for each instance), while in the current PyExtensionType this is a constant (a pickle of the class, which does not depend on the instance, only on the subclass). I suppose always calling it is not problematic here, also in the case where it would not be needed.